### PR TITLE
Add `AnimatableTableViewController` with designable `UIRefreshControl`

### PR DESCRIPTION
**Pull request by phimage**
and other features same as `AnimatableViewController`

`RefreshControlDesignable` allow to custom the internal `UIRefreshControl`

`UIRefreshControl` must not be set in subview but using refreshControl attribute of `UITableViewController`

ReStart of PR #418 due to many conflict
Modification since review by @tbaranes and @JakeLin 
- remove some space and add missing open on IBoutlet variable copied from AnimatableViewController
- add `TimeLineTableViewController` in demo app , to show how refresh control could be customised (simulate the refresh using `DispatchQueue`)
- fix a bug : if view not already loaded. If view not loaded, refreshControl could not be set

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -16,13 +16,15 @@ All notable changes to this project will be documented in this file.
 [#403](https://github.com/IBAnimatable/IBAnimatable/pull/403) by [@tbaranes](https://github.com/tbaranes)
 - Make images of `AnimatableSlider` designable.
 [#417](https://github.com/IBAnimatable/IBAnimatable/pull/417) by [@phimage](https://github.com/phimage)
-
+- Add `AnimatableTableViewController` to support `UIRefreshControl` customization.
+[#418](https://github.com/IBAnimatable/IBAnimatable/pull/418) by [@phimage](https://github.com/phimage)
+CHANGELOG
```
Comment (tbaranes) [inline CHANGELOG.md]: There's a `CHANGELOG` to remove here!

File: Documentation/APIs.md [hunk 2]
```diff
@@ -170,6 +171,26 @@ Easily add color layer on top of the UI element especially `AnimatableImageView`
 | ------------- |:-------------:| ----- |
 | hideNavigationBar | Bool | whether to hide navigation bar. The default value is `false`. |
 
+#### `RefreshControlDesignable`
+Display a `UIRefreshControl` in `AnimatableTableViewController`
```
Comment (tbaranes) [inline Documentation/APIs.md]: I wouldn't add this line, I think the protocol name is clear enough (moreover, we may conform to this protocol in other classes).

File: IBAnimatable/AnimatableTableViewController.swift [hunk 3]
```diff
@@ -0,0 +1,111 @@
+//
+//  AnimatableTableViewController.swift
```
Comment (tbaranes) [inline IBAnimatable/AnimatableTableViewController.swift]: You can remove these two lines. Take a look at the other headers of the project, we are just keeping the author lines.

File: IBAnimatable/RefreshControlerDesignable.swift [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+   The background color of the refresh control
+   */
+  var refreshControlBackgroundColor: UIColor? { get set }
+
+}
+
+public extension RefreshControlDesignable where Self: UITableViewController {
+
+  func configureRefreshController() {
+    if isViewLoaded {
```
Comment (tbaranes) [inline IBAnimatable/RefreshControlerDesignable.swift]: Guard?

File: IBAnimatableApp/TimeLineTableViewController.swift [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    self.refreshControl?.addTarget(self, action: #selector(TimeLineTableViewController.refresh(_:)), for: .valueChanged)
+  }
+
+  public func refresh(_ refreshControl: UIRefreshControl) {
+    // could update attributedTitle of refreshControl here
+
+    // Simulate an asynchrone refresh, could be a network request...
+    let time: TimeInterval = 5
+    self.updateMessage(refreshControl: refreshControl, time: time)
+    DispatchQueue.background.after(time) {
```
Comment (tbaranes) [inline IBAnimatableApp/TimeLineTableViewController.swift]: Why not directly running this on `main`?

File: IBAnimatableApp/TimeLineTableViewController.swift [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      // end refreshing, maybe reload table data if you do not implement table delegate to update each insert, update and delete events
+      DispatchQueue.main.async {
+        refreshControl.endRefreshing()
+      }
+    }
+  }
+
+  // Function to update message recursively
+  func updateMessage(refreshControl: UIRefreshControl, time: TimeInterval) {
+    if time < 0 {
```
Comment (tbaranes) [inline IBAnimatableApp/TimeLineTableViewController.swift]: guard?

**Review by tbaranes — CHANGES_REQUESTED** [review]
I let you a few more minor comments, if you can take a look at them... :)

Also, a new swiftlint warning appeared since we enabled `line_length`, can you take care of it? If you want to see the warnings directly in Xcode, you just have to install [SwiftLint](https://github.com/realm/SwiftLint)

**Comment by tbaranes** [comment]
One more thing that we should have precise it earlier, my apologies. We don't add features to the `Demo.storyboard`, but directly in the `Playground` part which is a showoff of all the features available in `IBAnimatable`.

For the refresh control, you could add an entry either in:

- `Playground` -> `User Interface` -> `Refresh Control`
- `Playground` -> `RefreshControl`  (new) -> `UITableView` 
- Third place?

Both cases are ok, it's up to you, maybe depending if we want to add an example for the `collectionView` (when we will support it)

Sorry for not letting you know about this before 😟

File: IBAnimatableApp/TimeLineTableViewController.swift (hunk 5 above)
Comment (phimage) [reply IBAnimatableApp/TimeLineTableViewController.swift]: To simulate a good practice, doing a  job in background like get data from network

**Comment by phimage** [comment]
I use swiftlint and update it
I have missing one

File: IBAnimatable/RefreshControlerDesignable.swift (hunk 4 above)
Comment (phimage) [reply IBAnimatable/RefreshControlerDesignable.swift]: agree

File: IBAnimatableApp/TimeLineTableViewController.swift (hunk 6 above)
Comment (phimage) [reply IBAnimatableApp/TimeLineTableViewController.swift]: ummm here not agree `if` is shorter and more understandable  but I will do it

**Comment by phimage** [comment]
I have no time to make new storyboards in playground 
And I have no internet at home currently, could merge without the demo?

**Review by tbaranes — APPROVED** [review]
Let's wait for @JakeLin review, but  I'm ok to merge it without the demo. We can add it later! Thanks for taking care of the comments.

File: IBAnimatable/AnimatableTableViewController.swift [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  @IBInspectable open var hideNavigationBar: Bool = false
+
+  // MARK: - StatusBarDesignable
+  @IBInspectable open var lightStatusBar: Bool = false
+
+  // MARK: - RootWindowDesignable
+  @IBInspectable open var rootWindowBackgroundColor: UIColor?
+
+  // MARK: - TransitionAnimatable
+  @IBInspectable open var _transitionAnimationType: String? {
```
Comment (JakeLin) [inline IBAnimatable/AnimatableTableViewController.swift]: I think we don't want that to be `open` and let it only visible in Interface Builder. We can remove `open` and let it as  the default `internal`.

File: IBAnimatable/AnimatableTableViewController.swift [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        transitionAnimationType = TransitionAnimationType(string: _transitionAnimationType)
+      }
+    }
+  }
+  open var transitionAnimationType: TransitionAnimationType = .none
+
+  @IBInspectable open var transitionDuration: Double = .nan
+
+  open var interactiveGestureType: InteractiveGestureType = .none
+  @IBInspectable open var _interactiveGestureType: String? {
```
Comment (JakeLin) [inline IBAnimatable/AnimatableTableViewController.swift]: Same comment as above.

File: IBAnimatableApp/RefreshTableViewController.swift [hunk 9]
```diff
@@ -0,0 +1,60 @@
+//
+//  Created by Eric Marchand on 31/03/2017.
+//  Copyright © 2017 IBAnimatable. All rights reserved.
+//
+
+import UIKit
+import IBAnimatable
+
+public class RefreshTableViewController: AnimatableTableViewController {
```
Comment (JakeLin) [inline IBAnimatableApp/RefreshTableViewController.swift:9]: Nice demo app 👍

**Review by JakeLin — APPROVED** [review]
👍 Well done @phimage

**Comment by JakeLin** [comment]
@phimage I have fixed my comment and merged the PR now 🎉🎉🎉. Thank you very much for creating such a great feature for IBAnimatable. It makes our life easier to configure the refersh control in Interface Builder.

For `@IBInspectable var _transitionAnimationType: String?`, because Interface Builder can't support enum. And we want to have nice programmatical API with strongly type like `open var transitionAnimationType: TransitionAnimationType = .none`. To do that, we found out a workaround, we use `String` as the type to make the inspectable property can be configured in Interface Builder. But don't want the user/client code to see `_transitionAnimationType`, then we make them as `internal`. In that case, we can configure `transitionAnimationType` in Interface Builder and use it in code as an enum. More info can be found in #218 @lastMove and @tbaranes did a great job on that.

**Comment by tbaranes** [comment]
For your information, I added an entry in #238 in order to not forget that we have to update the playground storyboard in order to have a demo for refresh control.

**Pull request merged**